### PR TITLE
IGNITE-23441 Sql. Cancellation of script execution

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-23441

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/QueryExecutor.java [hunk 1]
```diff
@@ -202,14 +202,16 @@ CompletableFuture<AsyncSqlCursor<InternalSqlRow>> executeChildQuery(
         }
 
         try {
-            trackQuery(query, null);
+            trackQuery(query, parent.cancellationToken);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/QueryExecutor.java]: Thanks. Reverted this change. 
From my point of view, it is enough to pass null for the child query.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionContext.java [hunk 2]
```diff
@@ -106,7 +116,17 @@ public CompletableFuture<Void> handleControlStatement(SqlNode node) {
...
-            this.wrapper = new ScriptTransactionWrapperImpl(tx, txTracker);
+            ScriptTransactionWrapperImpl wrapper0 = new ScriptTransactionWrapperImpl(tx, txTracker);
+            this.wrapper = wrapper0;
+
+            if (cancellationToken != null) {
+                // If the user cancels the script, we need to rollback the script transaction.
+                CancelHandleHelper.addCancelAction(
+                        cancellationToken,
+                        () -> wrapper0.rollback(new QueryCancelledException()),
+                        nullCompletedFuture()
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionContext.java]: Fixed.

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java [hunk 3]
```diff
@@ -558,4 +560,23 @@ public static void executeUpdate(String query, IgniteSql sql, @Nullable Transact
     public static void executeUpdate(String query, IgniteSql sql) {
         executeUpdate(query, sql, null);
     }
+
+    /** The action is expected to throw a {@link SqlException} with code {@link Sql#EXECUTION_CANCELLED_ERR}. */
+    public static void expectQueryCancelled(Executable action) {
+        assertThrowsSqlException(
+                Sql.EXECUTION_CANCELLED_ERR,
+                "The query was cancelled while executing.",
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java]: Fixed, thanks.

File: modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java [hunk 4]
```diff
@@ -558,4 +560,23 @@ public static void executeUpdate(String query, IgniteSql sql, @Nullable Transact
...
+        );
+    }
+
+    /** The action is expected to throw a {@link QueryCancelledException}. */
+    public static void expectQueryCancelledInternalException(Executable action) {
+        //noinspection ThrowableNotThrown
+        IgniteTestUtils.assertThrows(
+                QueryCancelledException.class,
+                action,
+                "The query was cancelled while executing."
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/testFixtures/java/org/apache/ignite/internal/sql/engine/util/SqlTestUtils.java]: Fixed, thanks.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java [hunk 5]
```diff
@@ -74,8 +75,23 @@ public void cancelQueryString() throws InterruptedException {
...
+
+        // Checks the exception that is thrown if a query is canceled before a cursor is obtained.
+        CancelHandle cancelHandle = CancelHandle.create();
+        CancellationToken token = cancelHandle.token();
+        cancelHandle.cancel();
+
+        //noinspection resource
+        assertThrowsSqlException(
+                Sql.EXECUTION_CANCELLED_ERR,
+                "The query was cancelled while executing.",
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java]: Fixed, thanks.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItCancelQueryTest.java [hunk 6]
```diff
@@ -18,30 +18,29 @@
...
-import org.apache.ignite.sql.SqlException;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 
 /** Set of test cases for query cancellation. */
-public class ItQueryCancelTest extends BaseSqlIntegrationTest {
+public class ItCancelQueryTest extends BaseSqlIntegrationTest {
+    private static final String QUERY_CANCELED_ERR = "The query was cancelled while executing";
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItCancelQueryTest.java]: Fixed, thanks.

File: modules/api/src/main/java/org/apache/ignite/sql/IgniteSql.java [hunk 7]
```diff
@@ -379,6 +379,16 @@ <T> CompletableFuture<AsyncResultSet<T>> executeAsync(
...
 
+    /**
+     * Executes a multi-statement SQL query.
+     *
+     * @param cancellationToken Cancellation token or {@code null}.
+     * @param query SQL query template.
+     * @param arguments Arguments for the template (optional).
+     * @throws SqlException If failed.
+     */
+    void executeScript(@Nullable CancellationToken cancellationToken, String query, @Nullable Object... arguments);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/api/src/main/java/org/apache/ignite/sql/IgniteSql.java:390]: As far as I understand, we use the practice of separating Nullable parameters from varargs.

For example:
```
    ResultSet<SqlRow> execute(
            @Nullable Transaction transaction,
            @Nullable CancellationToken cancellationToken,
            String query,
            @Nullable Object... arguments
    );
```

**Review by zstan — APPROVED** [review]

**Review by lowka — APPROVED** [review]

**Pull request merged**